src/plugins/content_management/server/core/types.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { RequestHandlerContext } from '@kbn/core-http-request-handler-context-server';
+import { ContentCrud } from './crud';
+
+/** Context that is sent to all storage instance methods */
+export interface StorageContext {
+  requestHandlerContext?: RequestHandlerContext;
+}
+
+export interface ContentStorage {
+  /** Get a single item */
+  get(ctx: StorageContext, id: string, options: unknown): Promise<any>;
```

Comment:
I think it should return a typed object where the content/data item itself is `unknown`, something like this:

```suggestion
  get(ctx: StorageContext, id: string, options: unknown): Promise<ContentItem>;
  
  interface ContentItem<D = unknown> {
    data: D;
    // Possibility to add more fields here in the future
  }
```

We could also introduce a concept of "common schema" or "well known fields":

```suggestion
  get(ctx: StorageContext, id: string, options: unknown): Promise<ContentItem>;
  
  interface ContentItem<D = unknown> {
    data: D;
    fields: CommonFields;
  }
  
  interface CommonFields {
    title?: string;
    description?: string;
    createAt?: number;
    updatedAt?: number;
    createdBy?: string;
    updatedBy?: string;
    tags?: string[];
  }
```

Reply:
Also, maybe worth differentiating between a content item as it is provided by the storage provider vs the one once it is already in the CM system. Something like *Storage Item* vs *Content Item*.

```ts
interface StorageItem<Data = unknown> {
  id: string;
  data: Data;
  fields: CommonFields;
  // Possibility to add more fields here.
}

interface ContentItem<Data = unknown> {
  type: string;
  storageItem: StorageItem<Data>;
  // Possibility to add more fields here from Storage Layer and Event Stream.
}
```

so when somebody registers a "get" method, they return us `StorageItem`s:

```ts
get(ctx, id, options): Promise<StorageItem>
```

then internally we can add more metadata to that *Storage Item* (type, maybe something from Content Enhancements, maybe something ephemeral):

```ts
const contentItem: ContentItem = {
  type: 'dashboard',
  storageItem,
  // more CM metadata ...
};
```

Reply:
So, a dashboard could look like this:

```ts
const dashboard: ContentItem<{panels: unknown}> = {
  type: 'dashboard',
  folder: '',
  isPinned: false,
  storageItem: {
    id: 'xyz',
    data: {
      panels: {},
    },
    fields: {
      title: 'eCommerce Dashboard',
      createdBy: 'kimchy',
    },
  },
};
```

Reply:
Then on the RPC layer it is wrapped in yet another object:

```ts
const itemResult: ItemResult = {
  item: dashboard,
  // Any RPC-specific fields, for example:
  fromCache: true,
};
```

Reply:
> Also, maybe worth differentiating between a content item as it is provided by the storage provider vs the one once it is already in the CM system.

Let's discuss it online, I think we are not talking about the same thing. Seems like you are talking about our future Search layer with common schema. I am just putting the initial crud foundation that should not require any modification on the content shape stored in the DB layer.

Reply:
> Is there a reason we use any as part of the public interface instead of unknown?

This is the interface to implement. If we use `unknown` then when actually implemented and we type the response with `{ title: string; }` then TS throws an error because `unknown` can't be the provided type.